src/dependency-manager/src/service-config/v1.ts

```diff
@@ -1,142 +1,366 @@
+        value.default = value.default || { valueFrom: value.valueFrom };
+        delete value.valueFrom;
+        delete value.value_from;
+      }
+
+      output[key] = plainToClass(ParameterDefinitionSpecV1, value);
     } else {
-      output[key] = { default: value };
+      output[key] = plainToClass(ParameterDefinitionSpecV1, {
+        default: value,
+      });
     }
   }
   return output;
-}
+};
 
-function transformVolumes(input: any) {
-  const output: any = {};
+const transformVolumes = (input?: Dictionary<string | Dictionary<any>>): Dictionary<ServiceVolumeV1> | undefined => {
+  if (!input) {
+    return undefined;
+  }
+
+  const output: Dictionary<ServiceVolumeV1> = {};
   for (const [key, value] of Object.entries(input)) {
-    output[key] = value instanceof Object ? value : { host_path: value };
+    output[key] = value instanceof Object
+      ? plainToClass(ServiceVolumeV1, value)
+      : plainToClass(ServiceVolumeV1, { host_path: value });
   }
   return output;
-}
+};
+
+const transformInterfaces = (input?: Dictionary<string | Dictionary<any>>): Dictionary<InterfaceSpecV1> | undefined => {
+  if (!input) {
+    return undefined;
+  }
 
-function transformInterfaces(input: any) {
-  const output: any = {};
+  const output: Dictionary<InterfaceSpecV1> = {};
   for (const [key, value] of Object.entries(input)) {
-    output[key] = value instanceof Object ? value : { port: value };
+    output[key] = typeof value === 'object'
+      ? plainToClass(InterfaceSpecV1, value)
+      : plainToClass(InterfaceSpecV1, { port: value });
   }
   return output;
+};
+
+class NotificationSpecV1 extends BaseSpec {
+  @IsString()
+  description!: string;
 }
 
-interface ServiceNotificationsV1 {
-  [notification_name: string]: {
-    description: string;
-  };
+class RestSubscriptionDataV1 extends BaseSpec {
+  @IsString()
+  uri!: string;
+
+  @IsOptional()
+  headers?: Dictionary<string>;
 }
 
-interface ServiceSubscriptionsV1 {
-  [service_name: string]: {
-    [event_name: string]: {
-      uri: string;
-      headers?: { [key: string]: string };
-    };
-  };
+class SubscriptionSpecV1 extends BaseSpec {
+  @IsString()
+  @IsIn(['rest', 'grpc'])
+  type!: string;
+
+  @Type(() => RestSubscriptionDataV1)
+  @ValidateIf(obj => obj.type === 'rest')
+  @IsInstance(RestSubscriptionDataV1)
+  data?: RestSubscriptionDataV1;
+
+  async validate(options?: ValidatorOptions) {
+    let errors = await super.validate(options);
+    errors = await validateNested(this, 'data', errors, options);
+    return errors;
+  }
 }
 
-class ServiceDatastoreV1 {
+class ServiceDatastoreV1 extends BaseSpec {
+  @IsOptional()
+  @IsString()
   host?: string;
+
+  @IsOptional()
+  @IsNumber()
   port?: number;
+
+  @IsOptional()
+  @IsString()
   image?: string;
+
   @Transform(value => (transformParameters(value)))
-  parameters: { [key: string]: ServiceParameterV1 } = {};
-}
+  @IsOptional()
+  parameters?: Dictionary<ParameterDefinitionSpecV1>;
```

Comment:
I'm a fan of the Dictionary type

Reply:
The other way was just ugly and we use it everywhere. Thought this would encourage us to use a type more often.